File: app/src/main/java/com/example/vijay/myapplication/ImageInspectorFragment.java
package com.example.vijay.myapplication;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.RelativeLayout;

import java.io.IOException;
import java.util.Random;

import static android.app.Activity.RESULT_OK;

/** Fragment containing the ImageInspectorToolbar and the ImageInspectorView */
public class ImageInspectorFragment extends Fragment {

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_image_inspector, container, false);

        ImageButton addButton = (ImageButton)view.findViewById(R.id.addButton);
        addButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                addButtonAction();
            }
        });

        final ImageButton cameraButton = (ImageButton)view.findViewById(R.id.cameraButton);
        cameraButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                cameraButtonAction();
            }
        });

        ImageButton detailsButton = (ImageButton)view.findViewById(R.id.detailsButton);
        detailsButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                detailsButtonAction();
            }
        });

        _imageInspectorView = (ImageInspectorView)view.findViewById(R.id.imageInspectorView);

        return view;
    }


    void addButtonAction() {
        Log.d("ImageInspectorFragment", "addButtonAction");

        Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
        intent.setType("image/*");

//        startActivityForResult(intent, REQUEST_CODE_GALLERY_IMAGE);
        _imageInspectorView.setBitmap(null);
    }

    void cameraButtonAction() {
        Log.d("ImageInspectorFragment", "cameraButtonAction");
    }

    void detailsButtonAction() {
        Log.d("ImageInspectorFragment", "detailsButtonAction");
    }


    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode != RESULT_OK) {
            return;
        }

        switch (requestCode) {
            case REQUEST_CODE_GALLERY_IMAGE:
                if (data != null) {
                    Uri imageURI = data.getData();
                    try {
                        Bitmap bitmap = MediaStore.Images.Media.getBitmap(getContext().getContentResolver(), imageURI);

                        ImageInspectorView imageInspectorView = (ImageInspectorView)getView().findViewById(R.id.imageInspectorView);
                        imageInspectorView.setBitmap(bitmap);


                    } catch (IOException e) {
                        Log.e(null, "unable to create bitmap");
                    }

                }

            default:
                break;
        }

    }



    /** Request code to use to fetch image from gallery (Used in startActivityForResult) */
    private final int REQUEST_CODE_GALLERY_IMAGE = 1;

    /** The ImageInspectorView instance this fragment manages */
    private ImageInspectorView _imageInspectorView = null;

}

File: imagecompare/src/main/java/com/vjc/imagecompare/view/ImageInspectorFragment.kt
package com.vjc.imagecompare.view

import android.app.Fragment
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.provider.MediaStore
import android.text.Layout
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageButton
import com.vjc.imagecompare.R
import com.vjc.imagecompare.extensions.center
import com.vjc.imagecompare.model.MetaData
import java.util.*

class ImageInspectorFragment constructor() : Fragment() {
    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        val view = inflater?.inflate(com.vjc.imagecompare.R.layout.fragment_image_inspector, container, false)


        view?.let {
            val random = Random();

            it.setBackgroundColor(this.context.getColor(R.color.theme))

            _imageInspectorView = it.findViewById<ImageInspectorView>(R.id.imageInspectorView)

            val addButton = it.findViewById<ImageButton>(R.id.addButton)
            addButton.setOnClickListener(object : View.OnClickListener {
                override fun onClick(v: View?) {
                    addButtonAction()
                }
            })

            val cameraButton = it.findViewById<ImageButton>(R.id.cameraButton)
            cameraButton.setOnClickListener(object : View.OnClickListener {
                override fun onClick(v: View?) {
                    cameraButtonAction()
                }
            })

            val detailsButton = it.findViewById<ImageButton>(R.id.detailsButton)
            detailsButton.setOnClickListener(object : View.OnClickListener {
                override fun onClick(v: View?) {
                    detailsButtonAction();
                }
            })



        }


        return view
    }

    fun addButtonAction() {
        _imageDetailsView?.close()

        val intent = Intent(Intent.ACTION_GET_CONTENT)
        intent.type = "image/*"

        this.startActivityForResult(intent, com.vjc.imagecompare.view._REQUEST_CODE_IMAGE_GALLERY)
    }

    fun cameraButtonAction() {
        _imageDetailsView?.close()
        val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)



        this.startActivityForResult(intent, com.vjc.imagecompare.view._REEQUEST_CODE_IMAGE_CAMERA)
    }

    fun detailsButtonAction() {
        println("detailsButtonAction")

        if (_imageDetailsView != null && _imageDetailsView!!.isShown) {
            this.hideImageDetailsView()
        } else {
            this.showImageDetailsView()
        }


    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        when(requestCode) {
            _REQUEST_CODE_IMAGE_GALLERY -> {
                _imageInspectorView?.bitmapUri = data?.data
                var metaData: List<MetaData>? = null
                data?.let {
                    val array = MetaData.metaDataArray(it.data, context)
                    metaData = array.asList()
                }
                _imageDetailsView?.metaData = metaData
            }

            _REEQUEST_CODE_IMAGE_CAMERA -> {

            }
        }
    }




    private var _imageInspectorView: ImageInspectorView? = null
    private var _imageDetailsView: ImageDetailsView? = null

    private fun showImageDetailsView() {

        if (this.view !is ViewGroup) {
            return
        }
        val viewGroup = this.view as ViewGroup

        if (_imageDetailsView == null) {
            _imageDetailsView = ImageDetailsView(this.context)
        }

        if (_imageInspectorView != null) {
            _imageDetailsView!!.showIn(_imageInspectorView!!)
        }

    }

    private fun hideImageDetailsView() {
        _imageDetailsView?.close()
    }
}

private const val _REQUEST_CODE_IMAGE_GALLERY = 1
private const val _REEQUEST_CODE_IMAGE_CAMERA = 2